src/pure/RealTimeQueue.mo

```diff
@@ -0,0 +1,1026 @@
+        let states6 = States.step(States.step(States.step(States.step(States.step(States.step(states))))));
+        #rebal(states6)
+      }
+    };
+    // if the queue is in the middle of a rebalancing process: push the element and advance the rebalancing process by 4 steps
+    // move back into the idle state if the rebalancing is done
+    case (#rebal((dir, big0, small0))) switch dir {
+      case (#right) {
+        let big = BigState.push(big0, element);
+        let states4 = States.step(States.step(States.step(States.step((#right, big, small0)))));
+        debug assert states4.0 == #right;
+        switch states4 {
+          case (_, #big2(#idle(_, big)), #small3(#idle(_, small))) {
+            debug assert idlesInvariant(big, small);
+            #idles(big, small)
+          };
+          case _ #rebal(states4)
+        }
+      };
+      case (#left) {
+        let small = SmallState.push(small0, element);
+        let states4 = States.step(States.step(States.step(States.step((#left, big0, small)))));
+        debug assert states4.0 == #left;
+        switch states4 {
+          case (_, #big2(#idle(_, big)), #small3(#idle(_, small))) {
+            debug assert idlesInvariant(small, big);
+            #idles(small, big) // swapped because dir=left
+          };
+          case _ #rebal(states4)
+        }
+      }
+    }
+  };
+
+  /// Insert a new element on the back end of a queue.
+  /// Returns the new queue with all the elements of `queue`, followed by `element` on the back.
+  ///
+  /// Example:
+  /// ```motoko include=import
+  /// persistent actor {
+  ///   let queue = Queue.pushBack(Queue.pushBack(Queue.empty<Nat>(), 1), 2);
+  ///   assert Queue.peekBack(queue) == ?2;
+  ///   assert Queue.size(queue) == 2;
+  /// }
+  /// ```
+  ///
+  /// Runtime: `O(1)` worst-case!
+  ///
+  /// Space: `O(1)` worst-case!
+  public func pushBack<T>(queue : Queue<T>, element : T) : Queue<T> = reverse(pushFront(reverse(queue), element));
+
+  /// Remove the element on the front end of a queue.
+  /// Returns `null` if `queue` is empty. Otherwise, it returns a pair of
+  /// the first element and a new queue that contains all the remaining elements of `queue`.
+  ///
+  /// Example:
+  /// ```motoko include=import
+  /// import Runtime "mo:base/Runtime";
+  ///
+  /// persistent actor {
+  ///   do {
+  ///     let initial = Queue.pushBack(Queue.pushBack(Queue.empty<Nat>(), 1), 2);
+  ///     let ?(frontElement, remainingQueue) = Queue.popFront(initial) else Runtime.trap "Empty queue impossible";
+  ///     assert frontElement == 1;
+  ///     assert Queue.size(remainingQueue) == 1;
+  ///   }
+  /// }
+  /// ```
+  ///
+  /// Runtime: `O(1)` worst-case!
+  ///
+  /// Space: `O(1)` worst-case!
+  public func popFront<T>(queue : Queue<T>) : ?(T, Queue<T>) = switch queue {
+    case (#empty) null;
+    case (#one(x)) ?(x, #empty);
+    case (#two(x, y)) ?(x, #one(y));
+    case (#three(x, y, z)) ?(x, #two(y, z));
```

Comment:
FTR I found that re-ordering these tests last helps a little (also in reverse and pushFront). See #306.